subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheBuildOptionsIntegrationTest.groovy

```diff
@@ -742,6 +743,47 @@ class ConfigurationCacheBuildOptionsIntegrationTest extends AbstractConfiguratio
         configurationCache.assertStateStored()
     }
 
+    @Issue("gradle/gradle#14465")
+    def "configuration is cacheable when providers are used in settings"() {
+
+        given:
+        settingsFile << """
+            providers.systemProperty("org.gradle.booleanProperty").forUseAtConfigurationTime().orElse(false).get()
+        """
+        configurationCacheRun "help", "-Dorg.gradle.booleanProperty=true"
+        def firstRunOutput = removeVfsLogOutput(result.normalizedOutput)
+            .replaceAll(/Calculating task graph as no configuration cache is available for tasks: help\n/, '')
+            .replaceAll(/Configuration cache entry stored.\n/, '')
+
+        when:
+        configurationCacheRun "help", "-Dorg.gradle.booleanProperty=true"
+        def secondRunOutput = removeVfsLogOutput(result.normalizedOutput)
+            .replaceAll(/Reusing configuration cache.\n/, '')
+            .replaceAll(/Configuration cache entry reused.\n/, '')
+
+        then:
+        firstRunOutput == secondRunOutput
+    }
+
+    @Issue("gradle/gradle#14465")
+    def "configuration cache is invalidated after property change when providers are used in settings"() {
+
+        given:
+        settingsFile << """
+            providers.systemProperty("org.gradle.booleanProperty").forUseAtConfigurationTime().orElse(false).get()
+        """
+        configurationCacheRun "help", "-Dorg.gradle.booleanProperty=true"
+
+        when:
+        configurationCacheRun "help", "-Dorg.gradle.booleanProperty=false"
+        def secondRunOutput = result.normalizedOutput
+        println(secondRunOutput)
+
+        then:
+        secondRunOutput.contains("Configuration cache entry stored.")
+        secondRunOutput.contains("Calculating task graph as configuration cache cannot be reused because system property 'org.gradle.booleanProperty' has changed.")
```

Comment:
```suggestion

        then:
        configurationCache.assertStateStored()
        output.contains("because system property 'org.gradle.booleanProperty' has changed.")
```